Save 'Folder as workspace' in session

mere-human opened the pull request:
Save 'Folder as Workspace' in session XML only if checkbox in a dialog is enabled.
Create a customizable file dialog to save a session.

Fix #9165

donho commented:
@mere-human 
[As I said](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9165#issuecomment-749249442), it could be a feature for some users, but the bug for the other.
So sorry but it won't be included.

donho commented on PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp:
```diff
@@ -1022,7 +1022,7 @@ void FileBrowser::getDirectoryStructure(const TCHAR *dir, const std::vector<gene
 }
 
-void FileBrowser::addRootFolder(generic_string rootFolderPath)
+void FileBrowser::addRootFolder(generic_string rootFolderPath, bool shouldExpand)
```
I don't see the point of adding this boolean. OTOH, you should using `latestSelectedItem` in `<FileBrowser latestSelectedItem="C:\xxx\yyy\\config.xml">` tag to restore the selection on the last selected file of last session is more meaningful.

donho reviewed (changes requested):
1. Remove `bool shouldExpand` from addRootFolder function
2. Instead of `<workspace>` tag, use 
```
<FileBrowser latestSelectedItem="c:\foo\bar.txt">
        <root foldername="C:\foo" />
</FileBrowser>
```
whole node structure (and naming) in `config.xml` 
3. Restore the selection on the last selected file.
4. Add an option in Save session dialog to enable this feature.

mere-human commented:
I fixed all the comments and implemented a file save dialog with a checkbox.

The only issue I see is when multi-instance is enabled, then after loading a session, "Folder as Workspace" has 0 width.
I cannot figure out why it happens. 
I know that in this case another instance is launched with command line arguments.
However, it doesn't happen when I launch Notepad++ with command line arguments like "-openSession C:\my-session.xml"
Does anybody has any ideas?
![Screenshot 2021-01-02 205919](https://user-images.githubusercontent.com/9664141/103464662-2d677c80-4d3e-11eb-926a-7f880c2810fa.png)
![Screenshot 2021-01-02 205937](https://user-images.githubusercontent.com/9664141/103464665-30fb0380-4d3e-11eb-9446-ae8093f87faf.png)

donho commented:
@mere-human 
> I fixed all the comments and implemented a file save dialog with a checkbox.

I'm quite confused. After applying your PR, I launched compiled binary and call save session:

![image](https://user-images.githubusercontent.com/90293/103603506-12983200-4f0f-11eb-9096-dcbb0d0d2f2e.png)

then in save session dialog:

![image](https://user-images.githubusercontent.com/90293/103603359-bcc38a00-4f0e-11eb-9f23-860f3974f5aa.png)

I don't see the checkbox. Are we talking the same dialog?

donho commented:
@mere-human 
Never mind. I didn't enable the Folder as Workspace. I see the checkbox now (Well played).

> The only issue I see is when multi-instance is enabled, then after loading a session, "Folder as Workspace" has 0 width.

I cannot reproduce 0 width of "Folder as Workspace" - could you provide the full instructions please?

mere-human commented:
@donho 
> I don't see the checkbox. Are we talking the same dialog?

It is intended. When "Folder as Workspace" is not visible, the checkbox isn't shown. Sorry for the confusion.
Will it make more sense to show a checkbox in a disabled state for that case? Like this:
![image](https://user-images.githubusercontent.com/9664141/103622729-c6b1b100-4f3f-11eb-8608-4b1b62b03e6a.png)


> I cannot reproduce 0 width of "Folder as Workspace" - could you provide the full instructions please?

1. Launch Notepad++.
2. Menu > Settings > Preferences > Multi-Instance > Select "Always in multi-instance" then close and restart the app.
3. Open/create some file so that it is not in default state.
4. Load a previously saved session that contains "Folder as Workspace".
5. Result: "Folder as Workspace" is shown (Menu > View > Folder as Workspace is checked) but it has 0 width.

sasumner commented:
>Will it make more sense to show a checkbox in a disabled state for that case?

I like this (as a general rule of thumb) as it lets a user know that a feature is available -- if conditions are correct.
Of course, sometimes it may leave user wondering about which conditions are needed. :-)

mere-human commented:
@sasumner 
> I like this (as a general rule of thumb) as it lets a user know that a feature is available -- if conditions are correct.

Ok, I can update this PR with that change.

> Of course, sometimes it may leave user wondering about which conditions are needed. :-)

It could be nice to show a tooltip when a cursor is hovered over the checkbox to clarify that.
Unfortunately, it seems like common file dialogs don't provide such an option (I couldn't find that).

donho commented:
@mere-human 
> Will it make more sense to show a checkbox in a disabled state for that case?

Yes, it's better IMO. "Apple-like" contextual menu looks dynamic, but it's confusing for users - people will ask themselves about what they have (or not) seem.

> 1. Launch Notepad++.
> 2. Menu > Settings > Preferences > Multi-Instance > Select "Always in multi-instance" then close and restart the app.
> 3. ...

OK, I can reproduce now. It's a bug. Could be link to [this bug-fix](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8005). So it's a regression in v7.9.2.

OTOH, loading session in multi-instance mode opens a new instance - which is a bug (to me) but it should be there since long time.

donho reviewed (changes requested):
1. Change copyright year/name/email
2. Respect Notepad++ coding style - add prefix "_" to the variable members of class.

sasumner commented on PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp:
```diff
@@ -0,0 +1,251 @@
+		if (pDialog)
+			return false; // Avoid double initizliation
+
+		HRESULT hr = CoCreateInstance(CLSID_FileSaveDialog,
```
Maybe when @mere-human is done tapping the dark side for this one, he can update the "ghetto" directory picker box (see https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8513#issuecomment-653489990 ) with the COM power?  :-)

mere-human replied:
> Is there no other way to use the dark side power (COM) for adding just a simple checkbox?

I wish there was another way. 
But it seems it is the only way we can get a customizable file dialog.
I didn't find any other way. 
Besides, Microsoft recommends using this approach in favor of a legacy function GetSaveFileName:
> _Starting with Windows Vista, the Open and Save As common dialog boxes have been superseded by the Common Item Dialog._
https://docs.microsoft.com/en-us/windows/win32/api/commdlg/nf-commdlg-getsavefilenamea

The alternative is designing a dialog from scratch which just _looks_ like the file dialog and adding a checkbox to that. 
But it is an overkill as for me.

mere-human commented on PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.h:
```diff
@@ -0,0 +1,55 @@
+	bool getCheckboxState() const;
+
+private:
+	class Impl;
```
I wanted to avoid polluting other source files with includes and defines in header.
I include <shobjidl.h> and redefine _WIN32_WINNT for IFileDialog to work which might conflict with the code in other files.

mere-human commented on PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp:
```diff
@@ -0,0 +1,251 @@
+			if (checkboxLabel && checkboxLabel[0] != '\0')
+			{
+				const BOOL isChecked = FALSE;
+				hr = pCustomize->AddCheckButton(IDC_FILE_CHECKBOX, checkboxLabel, isChecked);
```
There is not much placement control for this.
It seems, the layout depends on other controls and the order in which they are added to the dialog.
Also, the controls can be grouped to stick together when the dialog is resized.

mere-human commented on PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp:
```diff
@@ -0,0 +1,251 @@
+// This file is part of Notepad++ project
+// Copyright (C)2020 Don HO <don.h@free.fr>
```
Is it ok to put it this way?
`// Copyright (C) 2021 The Notepad++ Contributors.`

mere-human commented on PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp:
```diff
@@ -0,0 +1,251 @@
+	bool isCheckboxActive = true;
+	std::vector<std::pair<generic_string, generic_string>> filterSpec;
+	TCHAR fileName[MAX_PATH * 8];
+	int index = -1;
```
Ok, sure.
I just saw in some places that public members are named without `_` and thought that was acceptable.
Like this one:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Notepad_plus.h#L92
```
struct VisibleGUIConf final
{
	bool isPostIt = false;
	bool isFullScreen = false;
```

donho reviewed (changes requested):
I just discover 1 regression and1 bug:
1. filebroser's open state is not saved in the saved session file even checkbox has been checked.
2.  if session extension is defined (for example `.sess`), then even user chooses *.* on extension, `.sess`is always added. I expect no extension added if *.* is chosen.

mere-human commented:
@donho 
> 1. filebroser's open state is not saved in the saved session file even checkbox has been checked.

Could you, please, elaborate on this? What are the exact steps to reproduce the issue?

Current save session behavior for Folder as Workspace (FaW) is like this:
1. FaW isn't opened, checkbox is greyed-out -> FaW info is not saved in session
2. FaW is opened, checkbox isn't checked -> FaW info isn't saved in session
3. FaW is opened, checkbox is checked -> FaW info is saved in session

Load session behavior:
1. FaW isn't open, session without FaW info -> FaW isn't shown after loading
2. FaW isn't open, session with FaW info -> FaW is shown after loading
3. FaW is open, session without FaW info -> FaW is kept as is without changes
4. FaW is open, session with FaW info -> FaW is shown with new directories from session

donho commented:
> FaW is opened, checkbox is checked -> FaW info is saved in session

I recheck again (applying the PR and recompiling Notepad++), FaW info is not saved in session XML file.

mere-human commented:
> > FaW is opened, checkbox is checked -> FaW info is saved in session
> 
> I rechecked again (applying the PR and recompiling Notepad++), FaW info is not saved in session XML file.

Fixed. This should be fine now.

> 2. if session extension is defined (for example `.sess`), then even user chooses _._ on extension, `.sess`is always added. I expect no extension added if _._ is chosen.

I checked this out in a clean installation of Notepad++ 7.9.2 and the behavior is the same. 
When I type a file name without an extension and choose _*.*_ type, the file is saved using the extension from preferences.

This is set by `setDefExt()` call [here](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/NppIO.cpp#L2208) .
Then, it is set as value for the `lpstrDefExt` member of the `OPENFILENAME` structure.
MSDN doc for that member:

> _The default extension. GetOpenFileName and GetSaveFileName append this extension to the file name if the user fails to type an extension._ 

https://docs.microsoft.com/en-us/windows/win32/api/commdlg/ns-commdlg-openfilenamea

I follow the same behavior using `IFileDialog::SetDefaultExtension()`.
But we can change that by simply removing the call to `setDefExt()`.
Or add a checkbox in MISC preferences to let user decide whether or not he wants the extension to be appended.

mere-human commented:
I wonder if I am doing it wrong by using force push.
Should I address the review comments by adding new commits instead?

sasumner commented:
>I wonder if I am doing it wrong by using force push.
Should I address the review comments by adding new commits instead?

The contributing guidelines suggest one commit.
However, once reviewing starts, the "one commit" theory goes out the window. :-)
At that point, I just add commits so that, if desired, one can easily "track the evolution".
I don't know if that is what is right (desired) or not, but it has been working for me.

sasumner commented:
>Or add a checkbox in MISC preferences to let user decide whether or not he wants the extension to be appended.

I believe we are talking about an extension on a session file.  If so, my feeling is that it is hardly important enough to warrant a preference about it, and if one were made it would likely be confused with the already existing _SaveAs_ setting.
BTW, did you know that the extension applied/not-applied thing has been the most-debated thing in the history of Notepad++? :-)

donho commented:
@mere-human 
> I checked this out in a clean installation of Notepad++ 7.9.2 and the behavior is the same.

Indeed. So let's keep that way, until some users complaint.

donho commented:
@mere-human 
> I wonder if I am doing it wrong by using force push.

No need to force push next time, just commit & push it - I use `git rebase -i` to squeeze all the commits anyway.

donho commented:
@sasumner 
> The contributing guidelines suggest one commit.

True. I'll remove this rule.

sasumner commented:
> I'll remove this rule.

I would keep it.  (Just don't enforce it.)
It encourages new contributors to think harder about their original effort, if they think they won't get to revise it. :-)

sasumner commented:
Are we considering these dialogs with the different Win api for the normal Open and SaveAs dialogs as well?
If so, that's fine, except we need to really be careful about the "extension applied" behavior.
I wasn't kidding about that being the "most debated thing in N++ history".
FYI: There is an open PR ( #9328 ) on the existing SaveAs preference setting.

@mere-human I wasn't kidding about the other thing I mentioned earlier as well:  Please improve the "folder picker" for the _Find in Files_ "..." button (if you want to -- see the earlier referenced issue).

donho commented:
> FYI: There is an open PR ( #9328 ) on the existing SaveAs preference setting.

OK. It will be the next review.

> Please improve the "folder picker" for the Find in Files "..." button (if you want to -- see the earlier referenced issue).

I cannot see better option we can have - I won't be against if such better option exists.

mere-human commented:
> Are we considering these dialogs with the different Win api for the normal Open and SaveAs dialogs as well?
> If so, that's fine, except we need to really be careful about the "extension applied" behavior.

I had an idea that CustomFileDialog might be used for other open/save dialogs as well.
However, I thought that it would be safer to check the behavior in a smaller scope (session) and then decide if we want to use it widespread or not.
Also, it seemed to me a bit out of scope of the original issue.
Maybe, we need to create a new issue about "using new style CustomFileDialog for open/save operations".

> @mere-human I wasn't kidding about the other thing I mentioned earlier as well: Please improve the "folder picker" for the _Find in Files_ "..." button (if you want to -- see the earlier referenced issue).

Yes, I added it to my list of issues. I will take a look.
I cannot assign myself to issues yet:)

sasumner commented:
>I thought that it would be safer to check the behavior in a smaller scope (session) and then decide if we want to use it widespread or not.

That seems smart. :-)

> it seemed to me a bit out of scope of the original issue.

Yes, I was not suggesting it as part of this current issue (discussion in one issue often leads to a new issue creation on tangents).

>we need to create a new issue about "using new style CustomFileDialog for open/save operations".

As you said, let's wait as we check on "smaller scope" of existing change. :-)
Thank you for your efforts on this one, BTW.

>I cannot assign myself to issues yet

I will see if **_I_** can assign you to it. :-)

sasumner commented:
@mere-human 

I noticed that, as submitted in this PR, the new Save dialog for sessions does NOT accept a path with forward slashes in it.  While this is certainly not a requirement of this issue/PR, it would be nice if `/` were accepted by such dialogs.  Do you know if there is a way to make this type of dialog accept this?

mere-human commented:
@sasumner 

> I noticed that, as submitted in this PR, the new Save dialog for sessions does NOT accept a path with forward slashes in it. While this is certainly not a requirement of this issue/PR, it would be nice if `/` were accepted by such dialogs. Do you know if there is a way to make this type of dialog accept this?

I tried typing a path like `E:/Development/notepad-plus-plus` into the address edit box, and it works:
![image](https://user-images.githubusercontent.com/9664141/104057944-9eee7180-51fb-11eb-99c7-b36aab090ce2.png)
![image](https://user-images.githubusercontent.com/9664141/104057956-a1e96200-51fb-11eb-82c5-efa4af9e1c3b.png)

If you mean, typing a path like `E:/Development/notepad-plus-plus/1.xml` into the **File name** edit box, then it doesn't work indeed.
However, a path with backslashes typed into the same edit box works.
This case should be fixed. I think it can be done using events interface (IFileDialogEvents::OnFileOk):
https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bb776913(v=vs.85)#onfileok

sasumner commented:
>If you mean, typing a path like E:/Development/notepad-plus-plus/1.xml into the File name edit box, then it doesn't work indeed.

Yes, that's what I meant, although in my limited test I just tried `PowerEditor/testfile.txt` (basing it off your lower screenshot just above), not a complete path.

If forward slashes could be made to work here, perhaps we could finally eliminate support for "old style" dialog in the Preferences.

mere-human commented:
> > If you mean, typing a path like E:/Development/notepad-plus-plus/1.xml into the File name edit box, then it doesn't work indeed.
> 
> Yes, that's what I meant, although in my limited test I just tried `PowerEditor/testfile.txt` (basing it off your lower screenshot just above), not a complete path.
> 
> If forward slashes could be made to work here, perhaps we could finally eliminate support for "old style" dialog in the Preferences.

I see.
Could you please create an issue for that? If no, maybe I can create it later.

sasumner commented:
>If forward slashes could be made to work here, perhaps we could finally eliminate support for "old style" dialog in the Preferences.

Again, I am getting ahead of myself.  Clarify:  If the efforts from this PR are mainlined into the usual File Open/SaveAs operations.

>Could you please create an issue for that? If no, maybe I can create it later.

I will, later as well. :-)

donho commented:
> If forward slashes could be made to work here, perhaps we could finally eliminate support for "old style" dialog in the Preferences.

That's one of reasons that I did accept the dark side power - it seems to me COM is the only way to customize the standard dialog in Windows.

sasumner commented:
Issue created:  https://github.com/notepad-plus-plus/notepad-plus-plus/issues/9374 "Save/Load Session file dialogs don't accept forward slashes in 'File name:' box"

sasumner commented:
@mere-human I made another one for you to have fun with : #9375

mere-human commented:
> 
> > I wonder if I am doing it wrong by using force push.
> 
> No need to force push next time, just commit & push it - I use `git rebase -i` to squeeze all the commits anyway.

> > I'll remove this rule.
> 
> I would keep it. (Just don't enforce it.)
> It encourages new contributors to think harder about their original effort, if they think they won't get to revise it. :-)


Maybe, it would good to rephrase the contributing guidelines to be something like this? I can create a PR for that.

_Create a PR with a single commit to make the review process easier. Address the review change requests by pushing new commits. Do not force push. All the PR commits are squashed before merging._

sasumner commented:
>Create a PR with a single commit to make the review process easier. Address the review change requests by pushing new commits. Do not force push. All the PR commits are squashed before merging.

I like your phraseology. :-)

> I can create a PR for that.

I think it would be less effort if @donho just sees your text above and does a quick change for it himself.
However, I've done PRs for that kind of change in the past because open PRs are "more visible" to @donho than "comments".  :-)

BTW, my exact process is: create a branch with issue number at end of branch name, commit as many times as I want, squash commits into 1, submit PR, proceed from there...  I'm not sure if that is the _best_ workflow, but it seems to be working.

donho commented:
>> Create a PR with a single commit to make the review process easier. Address the review change requests by pushing new commits. Do not force push. All the PR commits are squashed before merging.

> BTW, my exact process is: create a branch with issue number at end of branch name, commit as many times as I want, squash commits into 1, submit PR, proceed from there... I'm not sure if that is the best workflow, but it seems to be working.

Make a PR (@sasumner  or @mere-human) please.

mere-human commented:
> Make a PR (@sasumner or @mere-human) please.

Done #9400
